include/circt/Dialect/Moore/MooreOps.td

```diff
@@ -230,15 +230,32 @@ def ReadLValueOp : MooreOp<"read_lvalue", [SameOperandsAndResultType]> {
   }];
 }
 
+def ReadOp : MooreOp<"read", [
+  TypesMatchWith<"input and result types match",
+    "result", "input", "RefType::get($_self)">
+]> {
+  let summary = "Like llvm.load";
+  let description = [{
+    //TODO:
+  }];
+  let arguments = (ins Arg<RefType, "", [MemRead]>:$input);
+  let results = (outs UnpackedType:$result);
+  let assemblyFormat = [{
+    $input attr-dict `:` type($result)
+  }];
+  let hasVerifier = 1;
```

Comment:
The `TypesMatchWith` you have added should already cover everything you have in your verifier: if you make RefType wrap around an `UnpackedType` instead of MLIR's generic `Type`, RefType will do the checking for UnpackedType for you. And the `TypesMatchWith` will do the checking that input's type is the result's type wrapped in RefType. Therefore the hand-written verifier should no longer be needed.

Reply:
I get it!